### PR TITLE
feature/onboarding add onboarding and user defaults storage

### PR DESCRIPTION
**Pull request by nurbekismagulov**

### CODE REVIEW

File: zhazam/Modules/Onboarding/Model/OnboardingPage.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+//  Copyright © 2020 crococoders. All rights reserved.
+//
+
+import Foundation
+
+enum OnboardingType {
+    case interfaceStyle
+    case finish
+    
+    static let all: [OnboardingType] = [.interfaceStyle, .finish]
```
Comment (rymbekov) [inline zhazam/Modules/Onboarding/Model/OnboardingPage.swift]: You can use CaseIterable protocol that automatically generates an array property of all cases in an enum

File: zhazam/Modules/Loading/LoadingViewController.swift [hunk 2]
```diff
@@ -41,6 +41,15 @@ final class LoadingViewController: UIViewController {
                           completion: nil)
     }
     
+    private func routeToOnboarding() {
+        let pages = OnboardingType.all.map { type -> OnboardingPage in
+            return OnboardingPage(type: type)
+        }
+        let viewController = OnboardingPageViewController(with: pages)
```
Comment (alimsanzhar) [inline zhazam/Modules/Loading/LoadingViewController.swift]: space

File: zhazam/Modules/Loading/LoadingViewController.swift [hunk 3]
```diff
@@ -70,3 +79,8 @@ final class LoadingViewController: UIViewController {
         percentageLabel.text = number < Constants.maxAvailablePercentage ? "0\(number)0." : "\(number)0."
     }
 }
+
+extension UserDefaultsStorage {
+    @UserDefaultsBacked(key: "isOnboardingCompleted", defaultValue: false)
+    static var isOnboardingCompleted: Bool
+}
```
Comment (alimsanzhar) [inline zhazam/Modules/Loading/LoadingViewController.swift]: Already used in InterfaceStyleConfigurationCommand, why do you need it here?

File: zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+extension OnboardingPageViewController: UIPageViewControllerDataSource {
+    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
+        return nil
+    }
+    
+    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
+        return nil
+    }
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift:27]: why are you implementing these methods if they return nil?

File: zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import UIKit
+
+class OnboardingPageViewController: UIPageViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.dataSource = self
+    }
+
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift:17]: no need for space

File: zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        pageControl.pageIndicatorTintColor = R.color.defaultGray()
+        pageControl.translatesAutoresizingMaskIntoConstraints = false
+        return pageControl
+    }()
+    
+    init(with pages: [OnboardingPage]) {
+        super.init(transitionStyle: .scroll,
+                   navigationOrientation: .horizontal,
+                   options: nil)
+        self.pages = pages
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift]: set pages value before super and there will be no need for making pages as var and make it optional

File: zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+//  Copyright © 2020 crococoders. All rights reserved.
+//
+
+import UIKit
+
+class OnboardingPageViewController: UIPageViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.dataSource = self
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift:15]: no need for self.

File: zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+//  Created by Nurbek Ismagulov on 4/19/20.
+//  Copyright © 2020 crococoders. All rights reserved.
+//
+
+import UIKit
+
+class OnboardingPageViewController: UIPageViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/Model/OnboardingPageViewController.swift:14]: let separate super and self calls (для наглядности чтобы видеть какая часть кода является твоей реализацией)

File: zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.dataSource = self
+        self.delegate = self
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift]: remove self and move it to separate method

File: zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    private func setupPageControl() {
+        view.addSubview(pageControl)
+        
+        let constraints = [
+            pageControl.bottomAnchor.constraint(equalTo: self.view.bottomAnchor, constant: -50),
+            pageControl.centerXAnchor.constraint(equalTo: self.view.centerXAnchor)
+        ]
+        NSLayoutConstraint.activate(constraints)
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift]: Write NSLayoutConstraint.activate([...]) like sundell does, no need to instantiate a variable constants

File: zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    private func getPage(for index: Int) -> OnboardingViewController? {
+        guard let pages = pages else { return nil }
+        guard index >= 0, index < pages.count else { return nil }
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift]: you can write it together

File: zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        return viewController
+    }
+}
+
+extension OnboardingPageViewController: UIPageViewControllerDataSource, UIPageViewControllerDelegate {
+    func pageViewController(_ pageViewController: UIPageViewController,
+                            viewControllerBefore viewController: UIViewController) -> UIViewController? {
+        guard let pageVC = viewController as? OnboardingViewController else { return nil }
+        let index = pageVC.getIndex()
+        guard index > 0 else { return nil }
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingPageViewController.swift]: here too

File: zhazam/Modules/Onboarding/View/OnboardingViewController.swift [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        super.init(nibName: String(describing: Self.self), bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        return nil
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configure()
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingViewController.swift]: separate

File: zhazam/Modules/Onboarding/View/OnboardingViewController.swift [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            stackView.addArrangedSubview(horizontalView)
+        case .finish:
+            let titledTextView = TitledTextView(title: R.string.localizable.letsStart(),
+                                                placeholder: R.string.localizable.start(),
+                                                completionString: R.string.localizable.start())
+            titledTextView.onCompleteBlock = { [weak self] in
+                UserDefaultsStorage.isOnboardingCompleted = true
+                self?.routeToMainScreen()
+            }
+            stackView.addArrangedSubview(titledTextView)
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingViewController.swift]: space

File: zhazam/Modules/Onboarding/View/OnboardingViewController.swift [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    
+    private func configure() {
+        switch page.type {
+        case .interfaceStyle:
+            let horizontalView = HorizontalKeyValueView(key: R.string.localizable.light(),
+                                                        value: getInterfaceStyleString())
+            horizontalView.onTapBlock = { [weak self] in
+                self?.changeInterfaceStyle()
+            }
+            stackView.addArrangedSubview(horizontalView)
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingViewController.swift]: space

File: zhazam/Modules/Onboarding/View/OnboardingViewController.swift [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    private func getInterfaceStyleString() -> String {
+        return darkModeEnabled() ? R.string.localizable.off() : R.string.localizable.on()
+    }
+    
+    private func darkModeEnabled() -> Bool {
+        guard #available(iOS 12.0, *),
+            self.traitCollection.userInterfaceStyle == .dark else {
+            return false
+        }
+        return true
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingViewController.swift]: space

File: zhazam/Modules/Onboarding/View/OnboardingViewController.swift [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    
+    private func changeInterfaceStyle() {
+        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else {
+           return
+        }
+        appDelegate.changeInterfaceStyle()
```
Comment (alimsanzhar) [inline zhazam/Modules/Onboarding/View/OnboardingViewController.swift]: space

File: zhazam/Modules/Views/HorizontalKeyValueView/HorizontalKeyValueView.swift [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+//  Copyright © 2020 crococoders. All rights reserved.
+//
+
+import UIKit
+
+final class HorizontalKeyValueView: UIView {
+    
+    @IBOutlet private var keyLabel: UILabel!
+    @IBOutlet private var valueLabel: UILabel!
+    var onTapBlock: (() -> Void)?
```
Comment (alimsanzhar) [inline zhazam/Modules/Views/HorizontalKeyValueView/HorizontalKeyValueView.swift]: space (everywhere)

Comment (alimsanzhar) [inline zhazam/Modules/Views/HorizontalKeyValueView/HorizontalKeyValueView.swift]: can use typealias for callback, create a global one so everyone can use it

File: zhazam/Modules/Views/TitledTextView/TitledTextView.swift [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+//
+
+import UIKit
+
+final class TitledTextView: UIView {
+
+    @IBOutlet private var titleLabel: UILabel!
+    @IBOutlet private var textField: UITextField!
+    private var completionString: String?
+    var onCompleteBlock: (() -> Void)?
```
Comment (alimsanzhar) [inline zhazam/Modules/Views/TitledTextView/TitledTextView.swift]: let it be onComplete, without block

File: zhazam/Modules/Views/TitledTextView/TitledTextView.swift [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+extension TitledTextView: UITextFieldDelegate {
+    func textField(_ textField: UITextField,
+                   shouldChangeCharactersIn range: NSRange,
+                   replacementString string: String) -> Bool {
+        let updatedString = (textField.text as NSString?)?.replacingCharacters(in: range, with: string)
+        if updatedString == completionString {
+            onCompleteBlock?()
+        }
+        return true
```
Comment (alimsanzhar) [inline zhazam/Modules/Views/TitledTextView/TitledTextView.swift:59]: space

File: zhazam/Resources/Strings/en.lproj/Localizable.strings [hunk 21]
```diff
@@ -7,3 +7,9 @@
 */
 
 "AppName" = "Zhazam";
+"Language" = "Language";
+"Light" = "Light";
+"On" = "on";
+"Off" = "off";
+"LetsStart" = "Let's start?";
+"Start" = "start";
```
Comment (alimsanzhar) [inline zhazam/Resources/Strings/en.lproj/Localizable.strings]: use uppercased namings and where it needs to be lowercased call "lowercased()" method for string

File: zhazam/Utility/Extensions/Optional+Extension.swift [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+
+import Foundation
+
+protocol AnyOptional {
+    var isNil: Bool { get }
+}
+
+extension Optional: AnyOptional {
+    var isNil: Bool { self == nil }
+}
```
Comment (alimsanzhar) [inline zhazam/Utility/Extensions/Optional+Extension.swift:17]: why do we need this?

**Review by alimsanzhar — CHANGES_REQUESTED** [review]

File: zhazam/Modules/Loading/LoadingViewController.swift (hunk 3 above)
Comment (nurbekismagulov) [reply zhazam/Modules/Loading/LoadingViewController.swift]: Different properties

File: zhazam/Utility/Extensions/Optional+Extension.swift (hunk 22 above)
Comment (nurbekismagulov) [reply zhazam/Utility/Extensions/Optional+Extension.swift:17]: Used in UserDefaultsBacked for optional cases

**Pull request closed without merging**